rm-datasource/src/main/java/io/seata/rm/datasource/undo/parser/ProtostuffUndoLogParser.java

```diff
@@ -0,0 +1,209 @@
+/*
+ *  Copyright 1999-2019 Seata.io Group.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+package io.seata.rm.datasource.undo.parser;
+
+import io.protostuff.Input;
+import io.protostuff.LinkedBuffer;
+import io.protostuff.Output;
+import io.protostuff.Pipe;
+import io.protostuff.ProtostuffIOUtil;
+import io.protostuff.Schema;
+import io.protostuff.WireFormat.FieldType;
+import io.protostuff.runtime.DefaultIdStrategy;
+import io.protostuff.runtime.Delegate;
+import io.protostuff.runtime.RuntimeEnv;
+import io.protostuff.runtime.RuntimeSchema;
+import io.seata.common.loader.LoadLevel;
+import io.seata.rm.datasource.undo.BranchUndoLog;
+import io.seata.rm.datasource.undo.UndoLogParser;
+
+import java.io.IOException;
+
+/**
+ * The type protostuff based undo log parser.
+ *
+ * @author Geng Zhang
+ */
+@LoadLevel(name = "protostuff")
+public class ProtostuffUndoLogParser implements UndoLogParser {
+
+    private final static DefaultIdStrategy ID_STRATEGY = ((DefaultIdStrategy) RuntimeEnv.ID_STRATEGY);
+
+    static {
+        ID_STRATEGY.registerDelegate(new DateDelegate());
+        ID_STRATEGY.registerDelegate(new TimestampDelegate());
+        ID_STRATEGY.registerDelegate(new SqlDateDelegate());
+        ID_STRATEGY.registerDelegate(new TimeDelegate());
+    }
+
+    private Schema<BranchUndoLog> schema = RuntimeSchema.getSchema(BranchUndoLog.class);
+
+    @Override
+    public String getName() {
+        return "protostuff";
+    }
+
+    @Override
+    public byte[] getDefaultContent() {
+        return new byte[0];
+    }
+
+    @Override
+    public byte[] encode(BranchUndoLog branchUndoLog) {
+        // Re-use (manage) this buffer to avoid allocating on every serialization
+        LinkedBuffer buffer = LinkedBuffer.allocate(512);
+        // ser
+        try {
+            return ProtostuffIOUtil.toByteArray(branchUndoLog, schema, buffer);
+        } finally {
+            buffer.clear();
+        }
+    }
+
+    @Override
+    public BranchUndoLog decode(byte[] bytes) {
+        if (bytes.length == 0) {
+            return new BranchUndoLog();
```

Comment:
return null？

Reply:
1. `FastjsonUndoLogParser` default content is `{}`, equivalent to `return new BranchUndoLog()`
2. If return null, we need to add more NPE check in `UndoLogManager`, such as :

```java
BranchUndoLog branchUndoLog = UndoLogParserFactory.getInstance().decode(rollbackInfo);
for (SQLUndoLog sqlUndoLog : branchUndoLog.getSqlUndoLogs()) {  <--- here

```